File: backend/internal/lexer/lexer.go
package lexer

import (
	"app/internal/token"
)

type InputType []rune

type Lexer struct {
	input   InputType
	pos     int  // Points to current
	readPos int  // Points to next character after current
	current rune // Current rune

	readContentIdx int
	content        []token.Token
}

func New(input string) *Lexer {
	lexer := Lexer{
		input:   []rune(input),
		pos:     0,
		readPos: 0,

		readContentIdx: 0,
		content:        make([]token.Token, 0),
	}
	lexer.tokenize()
	return &lexer
}

func (lexer *Lexer) Tokenize() []token.Token {
	return lexer.content
}

func (lexer *Lexer) tokenize() {
	lexer.readRune()

	tokens := make([]token.Token, 0)
	tok := lexer.readNextToken()
	for ; tok.Type != token.EOF; tok = lexer.readNextToken() {
		tokens = append(tokens, tok)
	}
	tokens = append(tokens, token.New(token.EOF, "EOF"))
	lexer.content = tokens
}

func (lexer *Lexer) readNextToken() token.Token {
	lexer.skipWhitespace()
	if lexer.pos >= len(lexer.input) {
		return token.New(token.EOF, "EOF")
	}

	var tok token.Token
	switch lexer.current {
	case '+':
		tok = token.New(token.PLUS, "+")
	case '-':
		tok = token.New(token.MINUS, "-")
	case '*':
		tok = token.New(token.ASTERISK, "*")
	case '/':
		tok = token.New(token.SLASH, "/")
	case '^':
		tok = token.New(token.CARET, "^")
	case '=':
		tok = token.New(token.EQUALS, "=")
	case '!':
		if lexer.peekRune() == '=' {
			tok = token.New(token.NOT_EQUALS, "!=")
			lexer.readRune()
		} else {
			tok = token.New(token.BANG, "!")
		}
	case '<':
		if lexer.peekRune() == '=' {
			tok = token.New(token.LE, "<=")
			lexer.readRune()
		} else {
			tok = token.New(token.LT, "<")
		}
	case '>':
		if lexer.peekRune() == '=' {
			tok = token.New(token.GE, ">=")
			lexer.readRune()
		} else {
			tok = token.New(token.GT, ">")
		}
	case ',':
		tok = token.New(token.COMMA, ",")
	case '(':
		tok = token.New(token.LPAREN, "(")
	case ')':
		tok = token.New(token.RPAREN, ")")
	case '\n':
		tok = token.New(token.NEW_LINE, "\\n")
	default:
		if isLetter(lexer.current) {
			literal := lexer.readPred(isLetter)
			tok = token.New(token.IDENT, literal)
			return tok
		} else if isNumber(lexer.current) {
			literal := lexer.readPred(numberPred())
			tok = token.New(token.NUMBER, literal)
			return tok
		} else {
			tok = token.New(token.ILLEGAL, string(lexer.current))
		}
	}

	lexer.readRune()
	return tok
}

func (lexer *Lexer) skipWhitespace() {
	for isWhitespace(lexer.current) {
		lexer.readRune()
	}
}

func isWhitespace(r rune) bool {
	return r == '\t' || r == '\r' || r == ' '
}

func isLetter(r rune) bool {
	return ('a' <= r && r <= 'z') || ('A' <= r && r <= 'Z') || r == '_'
}

func isNumber(r rune) bool {
	return ('0' <= r && r <= '9')
}

func numberPred() func(rune) bool {
	seenDot := false
	return func(r rune) bool {
		if isNumber(r) {
			return true
		}
		if (r == '.') && !seenDot {
			seenDot = true
			return true
		}
		return false
	}
}

func (lexer *Lexer) peekRune() rune {
	if lexer.readPos >= len(lexer.input) {
		return 0
	}
	return lexer.input[lexer.readPos]
}

func (lexer *Lexer) readPred(pred func(rune) bool) string {
	start := lexer.pos
	for pred(lexer.current) {
		lexer.readRune()
	}
	literal := lexer.input[start:lexer.pos]
	return string(literal)
}

func (lexer *Lexer) readRune() {
	if lexer.readPos >= len(lexer.input) {
		lexer.current = 0
	} else {
		lexer.current = lexer.input[lexer.readPos]
	}
	lexer.pos = lexer.readPos
	lexer.readPos += 1
}

File: backend/internal/lexing/lexer.py
from typing import Callable
from backend.internal.tokens import Token, TokenType

NULL_CODE: str = "\0"


class Lexer:
    def __init__(self, input: str) -> None:
        self._input = input

        self._pos: int = 0
        self._read_pos: int = 0
        self._current: str = ""

        self._read_content_idx: int = 0
        self._content: list[Token] = []

        self._tokenize()

    def tokenize(self) -> list[Token]:
        return self._content

    def _tokenize(self) -> None:
        self._content.clear()
        self._read_unicode()

        token = self._read_next_token()
        while token.ttype is not TokenType.EOF:
            self._content.append(token)
            token = self._read_next_token()
        self._content.append(Token(TokenType.EOF, "EOF"))

    def _read_unicode(self) -> None:
        self._current = (
            NULL_CODE
            if self._read_pos >= len(self._input)
            else self._input[self._read_pos]
        )
        self._pos = self._read_pos
        self._read_pos += 1

    def _read_next_token(self) -> Token:
        self._skip_whitespace()
        if self._pos >= len(self._input):
            return Token(TokenType.EOF, "EOF")

        token = Token(TokenType.EOF, "EOF")
        match self._current:
            case "+":
                token = Token(TokenType.PLUS, "+")
            case "-":
                token = Token(TokenType.MINUS, "-")
            case "*":
                token = Token(TokenType.ASTERISK, "*")
            case "/":
                token = Token(TokenType.SLASH, "/")
            case "^":
                token = Token(TokenType.CARET, "^")
            case "=":
                token = Token(TokenType.EQUALS, "=")
            case ",":
                token = Token(TokenType.COMMA, ",")
            case "(":
                token = Token(TokenType.LPAREN, "(")
            case ")":
                token = Token(TokenType.RPAREN, ")")
            case "\n":
                token = Token(TokenType.NEW_LINE, "\\n")
            case "!":
                if self._peek_unicode() == "=":
                    token = Token(TokenType.NOT_EQUALS, "!=")
                    self._read_unicode()
                else:
                    token = Token(TokenType.BANG, "!")
            case "<":
                if self._peek_unicode() == "=":
                    token = Token(TokenType.LE, "<=")
                    self._read_unicode()
                else:
                    token = Token(TokenType.LT, "<")
            case ">":
                if self._peek_unicode() == "=":
                    token = Token(TokenType.GE, ">=")
                    self._read_unicode()
                else:
                    token = Token(TokenType.GT, ">")
            case _:
                if is_letter(self._current):
                    literal = self._read_pred(is_letter)
                    return Token(TokenType.IDENT, literal)
                elif is_number(self._current):
                    literal = self._read_pred(number_pred())
                    return Token(TokenType.NUMBER, literal)
                else:
                    token = Token(TokenType.ILLEGAL, self._current)

        self._read_unicode()
        return token

    def _skip_whitespace(self) -> None:
        while is_whitespace(self._current):
            self._read_unicode()

    def _peek_unicode(self) -> str:
        if self._read_pos >= len(self._input):
            return NULL_CODE
        return self._input[self._read_pos]

    def _read_pred(self, pred: Callable[[str], bool]) -> str:
        start = self._pos
        while pred(self._current):
            self._read_unicode()
        return self._input[start : self._pos]


def is_whitespace(code: str) -> bool:
    whitespaces: set[str] = {" ", "\r", "\t"}
    return code in whitespaces


def is_letter(code: str) -> bool:
    c = code.lower()
    return "a" <= c and c <= "z" or code == "_"


def is_number(code: str) -> bool:
    return "0" <= code and code <= "9"


def number_pred() -> Callable[[str], bool]:
    seen_dot = False

    def inner(code: str) -> bool:
        nonlocal seen_dot
        if is_number(code):
            return True
        elif (code == ".") and not seen_dot:
            seen_dot = True
            return True
        return False

    return inner
